### PR TITLE
fix: stack trace point to incorrect file (#3004)

### PR DESCRIPTION
**Pull request by ChenKS12138**
# Motivation

- Fixes https://github.com/vitest-dev/vitest/issues/3004

# Test Result

## Before Fix

image

## After Fix

image

### CODE REVIEW

**Comment by sheremet-va** [comment]
Stack should already be correct at this stage when we process the error.

**Comment by ChenKS12138** [comment]
> Stack should already be correct at this stage when we process the error.

It this means that `error` which is the params of function `printError` should be corrected before calling function `parseErrorStacktrace`?

Making `stackStr` to be something like  `Module.testFunc (/home/cattchen/codes/github.com/DanHulton/vitest-stack-sourcemap-repro/src/main.js:2:3)`  or  `Module.testFunc (src/main.js:2:3)` , instead of `Module.testFunc (/src/main.js:2:3)`

image

**Comment by sheremet-va** [comment]
> > Stack should already be correct at this stage when we process the error.
> 
> It this means that `error` which is the params of function `printError` should be corrected before calling function `parseErrorStacktrace`?
> 
> Making `stackStr` to be something like `Module.testFunc (/home/cattchen/codes/github.com/DanHulton/vitest-stack-sourcemap-repro/src/main.js:2:3)` or `Module.testFunc (src/main.js:2:3)` , instead of `Module.testFunc (/src/main.js:2:3)`
> 
> image
> 
> # UPDATE:
> It seems that stack has already to be correct at this stage when we process the error in somehow. `/src/main.js` is a valid path for `vite`, but not valid for `node:fs`, it depends on the point of view.
> 
> And We could barely change the `[[FunctionLocation]]` of a variable, which is used to create error stackframe.
> 
> image

Vitest gives full path to modules when processing files, so it should be a fullpath already. If it isn’t, then something is broken sonewhera. This error object is given to other custom reporters, so it should have correct stack trace before that.

**Comment by ChenKS12138** [comment]
> > > Stack should already be correct at this stage when we process the error.
> > 
> > 
> > It this means that `error` which is the params of function `printError` should be corrected before calling function `parseErrorStacktrace`?
> > Making `stackStr` to be something like `Module.testFunc (/home/cattchen/codes/github.com/DanHulton/vitest-stack-sourcemap-repro/src/main.js:2:3)` or `Module.testFunc (src/main.js:2:3)` , instead of `Module.testFunc (/src/main.js:2:3)`
> > image
> > # UPDATE:
> > It seems that stack has already to be correct at this stage when we process the error in somehow. `/src/main.js` is a valid path for `vite`, but not valid for `node:fs`, it depends on the point of view.
> > And We could barely change the `[[FunctionLocation]]` of a variable, which is used to create error stackframe.
> > image
> 
> Vitest gives full path to modules when processing files, so it should be a fullpath already. If it isn’t, then something is broken sonewhera. This error object is given to other custom reporters, so it should have correct stack trace before we print it.

Thanks for your tips! And it seems that function `originalPositionFor` returns the wrong `source`, and it finally leads to prepare wrong data to `Error.prepareStackTrace`.

6f28add6891676a6fdff800adacf11f

File: packages/vite-node/src/source-map-handler.ts [hunk 1]
```diff
@@ -192,8 +192,10 @@ function mapSourcePosition(position: OriginalMapping) {
     // better to give a precise location in the compiled file than a vague
     // location in the original file.
     if (originalPosition.source !== null) {
+      // `originalPositionFor` do not return the expected file source. So use `position.source`
+      // instead of `originalPosition.source`
       originalPosition.source = supportRelativeURL(
-        sourceMap.url, originalPosition.source)
+        sourceMap.url, position.source)
```
Comment (sheremet-va) [inline packages/vite-node/src/source-map-handler.ts:198]: Doesn't look like this is the fix. `originalPositionFor` should return the correct position. Is something wrong with the map?

**Review by sheremet-va — CHANGES_REQUESTED** [review]

File: packages/vite-node/src/source-map-handler.ts (hunk 1 above)
Comment (ChenKS12138) [reply packages/vite-node/src/source-map-handler.ts:198]: 1. function `originalPositionFor` works well, but the arguemnt `sourceMap.map` is broken, thus `originalPositionFor ` doesn't return the correct position. It seems that `ViteDevServer` doesn't return the `sources` in absolute path (`/home/cattchen/codes/github.com/DanHulton/vitest-stack-sourcemap-repro/src/main.js` is expected, but finally got `/src/main.js`). I suppose that it's better to use `position.source` than `originalPosition.source`

image

image

2. `sourceMap.url` and `supportRelativeURL` seems work well.

Comment (sheremet-va) [reply packages/vite-node/src/source-map-handler.ts:198]: Let's patch map's `sources` in `packages/vite-node/src/server.ts:195` then. I think `toFilePath` should return the correct path.

Comment (ChenKS12138) [reply packages/vite-node/src/source-map-handler.ts]: Thanks for your suggestions! It seems to work well.

image

image

**Comment by sheremet-va** [comment]
Can you add a test case that fails without this patch, but succeeds when the patch is applied?

**Comment by sheremet-va** [comment]
> I have added a test case, and checkout to commits includes `before patch`([d37fa3c](https://github.com/vitest-dev/vitest/commit/d37fa3c75a341db7d135da1d1da9cee7ed9c65b4)) and `after patch`([35a6ef4](https://github.com/vitest-dev/vitest/commit/35a6ef46abd4216efcb2ad024c38676f087be042)) to run test, it works well.
> 
> ```shell
> pnpm build && git rev-parse HEAD && pnpm --filter='./test/vite-node' run test
> ```
> 
> 929eaab2dd80fef1f7094f7c2371f87
> 
> fba969bd40ca31917fa3929a087152a

Yes, but this test doesn’t really test the behavior. Can you add a test where it fails and prints correct stack trace? You can create these conditions in test/stacktrace.

**Comment by ChenKS12138** [comment]
Oh, I see. I have added a new test case in `test/stacktrace`, and it works.

```shell
pnpm build && git rev-parse HEAD && pnpm --filter='./test/stacktraces' run test
```

b8fcc19d7c01430276bf18010974572

1584c99ddcdbc99a78877b1132a63f3

**Comment by sheremet-va** [comment]
Thank you for fixing this!

**Pull request merged**